File: innerlibs1/Web.vb
Imports System.Collections.Specialized
Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Runtime.CompilerServices
Imports System.Text
Imports System.Text.RegularExpressions

Imports System.Xml



''' <summary>
''' Modulo Web
''' </summary>
''' <remarks></remarks>
Public Module Web

    Public Iterator Function GetLocalIP() As IEnumerable(Of String)

        Dim host = Dns.GetHostEntry(Dns.GetHostName())
        For Each ip In host.AddressList
            If (ip.AddressFamily = AddressFamily.InterNetwork) Then
                Yield ip.ToString()
            End If
        Next


    End Function


    <Extension> Public Function ParseQueryString(ByVal URL As Uri) As NameValueCollection
        Return URL.Query.ParseQueryString()
    End Function

    ''' <summary>
    ''' Retorna o Titulo do arquivo a partir do nome do arquivo
    ''' </summary>
    ''' <param name="Info"> Arquivo ou Diretório</param>
    ''' <returns></returns>
    <Extension()> Function FileNameAsTitle(Info As FileSystemInfo) As String
        Return Path.GetFileNameWithoutExtension(Info.Name).CamelAdjust().Replace("_", " ").ToTitle
    End Function

    ''' <summary>
    ''' Retorna o Titulo do arquivo a partir do nome do arquivo
    ''' </summary>
    ''' <param name="FileName"> Arquivo ou Diretório</param>
    ''' <returns></returns>
    <Extension()> Function FileNameAsTitle(FileName As String) As String
        Return Path.GetFileNameWithoutExtension(FileName).CamelAdjust().Replace("_", " ").ToTitle
    End Function

    ''' <summary>
    ''' Minifica uma folha de estilo CSS
    ''' </summary>
    ''' <param name="CSS">String contendo o CSS</param>
    ''' <returns></returns>
    <Extension()> Public Function MinifyCSS(CSS As String) As String
        If CSS.IsNotBlank() Then
            CSS = Regex.Replace(CSS, "[a-zA-Z]+#", "#")
            CSS = Regex.Replace(CSS, "[\n\r]+\s*", String.Empty)
            CSS = Regex.Replace(CSS, "\s+", " ")
            CSS = Regex.Replace(CSS, "\s?([:,;{}])\s?", "$1")
            CSS = CSS.Replace(";}", "}")
            CSS = Regex.Replace(CSS, "([\s:]0)(px|pt|%|em)", "$1")
            ' Remove comments from CSS
            CSS = Regex.Replace(CSS, "/\*[\d\D]*?\*/", String.Empty)
        End If
        Return CSS
    End Function

    ''' <summary>
    ''' Minifica um arquivo JavaScript
    ''' </summary>
    ''' <param name="Js">String contendo o Javascript</param>
    ''' <returns></returns>
    <Extension()> Public Function MinifyJS(Js As String) As String
        Return New JSMin().Minify(Js)
    End Function

    ''' <summary>
    ''' Verifica se o computador está conectado com a internet
    ''' </summary>
    ''' <returns></returns>
    Public Function IsConnected(Optional Test As String = "http://google.com") As Boolean
        Try
            Using client = New WebClient()
                Using stream = client.OpenRead(Test)
                    Return True
                End Using
            End Using
        Catch
            Return False
        End Try
    End Function





    ''' <summary>
    ''' Adciona um parametro a Query String de uma URL
    ''' </summary>
    ''' <param name="Url">  Uri</param>
    ''' <param name="Key">  Nome do parâmetro</param>
    ''' <param name="Values">Valor do Parâmetro</param>
    ''' <returns></returns>
    <Extension>
    Public Function AddParameter(Url As Uri, Key As String, Append As Boolean, ParamArray Values As String()) As Uri
        Dim UriBuilder = New UriBuilder(Url)
        Dim query = UriBuilder.Query.ParseQueryString()
        If Values Is Nothing OrElse Append = False Then
            If query.AllKeys.Contains(Key) Then
                query.Remove(Key)
            End If
        End If
        For Each v In If(Values, {})
            query.Add(Key, v)
        Next
        UriBuilder.Query = query.ToString()
        Url = New Uri(UriBuilder.ToString())
        Return Url
    End Function

    ''' <summary>
    ''' Adciona um parametro a Query String de uma URL
    ''' </summary>
    ''' <param name="Url">  Uri</param>
    ''' <param name="Key">  Nome do parâmetro</param>
    ''' <param name="Values">Valor do Parâmetro</param>
    ''' <returns></returns>
    <Extension>
    Public Function AddParameter(Url As Uri, Key As String, ParamArray Values As String()) As Uri
        Return Url.AddParameter(Key, True, Values)
    End Function

    ''' <summary>
    ''' Adciona um parametro a Query String de uma URL
    ''' </summary>
    ''' <param name="Url">  Uri</param>
    ''' <param name="Key">  Nome do parâmetro</param>
    ''' <param name="Values">Valor do Parâmetro</param>
    ''' <returns></returns>
    <Extension()> Public Function RemoveParameter(Url As Uri, ParamArray Keys As String()) As Uri
        Dim UriBuilder = New UriBuilder(Url)
        Dim query = (UriBuilder.Query).ParseQueryString
        Keys = If(Keys IsNot Nothing AndAlso Keys.Count > 0, Keys, query.AllKeys)
        For Each k In Keys
            Try
                query.Remove(k)
            Catch ex As Exception
            End Try
        Next
        UriBuilder.Query = query.ToString
        Return UriBuilder.Uri
    End Function


    ''' <summary>
    ''' Retorna os segmentos de uma url
    ''' </summary>
    ''' <param name="Url"></param>
    ''' <returns></returns>
    <Extension()> Public Function GetUrlSegments(Url As String) As IEnumerable(Of String)

        Dim l As New List(Of String)
        Dim p As New Regex("(?<!\?.+)(?<=\/)[\w-.]+(?=[/\r\n?]|$)", RegexOptions.Singleline + RegexOptions.IgnoreCase)
        Dim gs = p.Matches(Url)
        For Each g As Match In gs
            l.Add(g.Value)
        Next
        Return l

    End Function

    ''' <summary>
    ''' Substitui os parametros de rota de uma URL por valores de um objeto
    ''' </summary>
    ''' <typeparam name="T"></typeparam>
    ''' <param name="obj"></param>
    ''' <param name="UrlPattern"></param>
    ''' <returns></returns>
    <Extension()> Public Function ReplaceUrlParameters(Of T)(UrlPattern As String, obj As T) As String
        UrlPattern = Regex.Replace(UrlPattern, "{([^:]+)\s*:\s*(.+?)(?<!\\)}", "{$1}")
        If obj IsNot Nothing Then UrlPattern = UrlPattern.Inject(obj)
        Return UrlPattern.RemoveLastEqual("/")
    End Function

    <Extension()> Public Function RemoveUrlParameters(UrlPattern As String) As String
        UrlPattern = Regex.Replace(UrlPattern, "{([^:]+)\s*:\s*(.+?)(?<!\\)}", "")
        Return UrlPattern.RemoveLastEqual("/")
    End Function



    ''' <summary>
    ''' Captura o Username ou UserID de uma URL do Facebook
    ''' </summary>
    ''' <param name="URL">URL do Facebook</param>
    ''' <returns></returns>
    <Extension> Public Function GetFacebookUsername(URL As String) As String
        If URL.IsURL AndAlso URL.GetDomain.ToLower.IsAny("facebook.com", "fb.com") Then
            Return Regex.Match(URL, "(?:(?:http|https):\/\/)?(?:www.)?facebook.com\/(?:(?:\w)*#!\/)?(?:pages\/)?(?:[?\w\-]*\/)?(?:profile.php\?id=(?=\d.*))?([\w\-]*)?").Groups(1).Value
        Else
            Throw New Exception("Invalid Facebook URL")
        End If
    End Function

    ''' <summary>
    ''' Captura o Username ou UserID de uma URL do Facebook
    ''' </summary>
    ''' <param name="URL">URL do Facebook</param>
    ''' <returns></returns>
    <Extension> Public Function GetFacebookUsername(URL As Uri) As String
        Return GetFacebookUsername(URL.AbsoluteUri)
    End Function

    ''' <summary>
    ''' Captura a Thumbnail de um video do youtube
    ''' </summary>
    ''' <param name="URL">Url do Youtube</param>
    ''' <returns></returns>
    Public Function GetYoutubeThumbnail(URL As String) As Byte()
        Return GetFile("http://img.youtube.com/vi/" & GetVideoId(URL) & "/hqdefault.jpg")
    End Function

    Public Function GetString(URL) As String
        Dim s = ""
        Using c As New WebClient
            s = c.DownloadString(URL)
        End Using
        Return s
    End Function
    Public Function GetFile(URL As String) As Byte()
        Dim s As Byte()
        Using c As New WebClient
            s = c.DownloadData(URL)
        End Using
        Return s
    End Function

    Public Function GetImage(URL As String) As System.Drawing.Image
        Return GetFile(URL).ToImage()
    End Function


    ''' <summary>
    ''' Captura a Thumbnail de um video do youtube
    ''' </summary>
    ''' <param name="URL">Url do Youtube</param>
    ''' <returns></returns>
    Public Function GetYoutubeThumbnail(URL As Uri) As Byte()
        Return GetYoutubeThumbnail(URL.AbsoluteUri)
    End Function

    ''' <summary>
    ''' Captura o ID de um video do YOUTUBE ou VIMEO em uma URL
    ''' </summary>
    ''' <param name="URL">URL do video</param>
    ''' <returns>ID do video do youtube ou Vimeo</returns>

    Public Function GetVideoId(URL As String) As String
        If URL.IsURL Then
            Select Case True
                Case URL.GetDomain.ContainsAny("youtube", "youtu")
                    Return Regex.Match(URL.ReplaceNone("&feature=youtu.be"), "(?:https?:\/\/)?(?:www\.)?youtu(?:.be\/|be\.com\/watch\?v=|be\.com\/v\/)(.{8,})").Groups(1).Value
                Case URL.GetDomain.ContainsAny("vimeo")
                    Return Regex.Match(URL, "vimeo\.com/(?:.*#|.*/videos/)?([0-9]+)").Groups(1).Value
                Case Else
                    Throw New Exception("Invalid Youtube or Vimeo URL")
            End Select
        Else
            Throw New Exception("Invalid Youtube or Vimeo URL")
        End If
    End Function

    ''' <summary>
    ''' Captura o ID de um video do youtube em uma URL
    ''' </summary>
    ''' <param name="URL">URL do video</param>
    ''' <returns>ID do video do youtube</returns>
    <Extension>
    Public Function GetVideoId(URL As Uri) As String
        Return GetVideoId(URL.AbsoluteUri)
    End Function

    ''' <summary>
    ''' Verifica se um site está indisponível usando o serviço IsUp.Me
    ''' </summary>
    ''' <param name="Url">Url</param>
    ''' <returns>True para site fora do Ar</returns>

    <Extension>
    Public Function IsDown(Url As String) As Boolean
        Dim content As String = GetString("http://downforeveryoneorjustme.com/" & Url)
        If content.Contains("It's just you") Then
            Return False
        Else
            Return True
        End If
    End Function

    ''' <summary>
    ''' Verifica se um site está disponível usando o serviço IsUp.Me
    ''' </summary>
    ''' <param name="Url">Url</param>
    ''' <returns>False para site fora do Ar</returns>

    <Extension>
    Public Function IsUp(Url As String) As Boolean
        Return Not Url.IsDown()
    End Function

    ''' <summary>
    ''' Verifica se um site está indisponível usando o serviço IsUp.Me
    ''' </summary>
    ''' <param name="Url">Url</param>
    ''' <returns>True para site fora do Ar</returns>

    Public Function IsDown(Url As Uri) As Boolean
        Return Url.AbsoluteUri.IsDown()
    End Function

    ''' <summary>
    ''' Verifica se um site está disponível usando o serviço IsUp.Me
    ''' </summary>
    ''' <param name="Url">Url</param>
    ''' <returns>False para site fora do Ar</returns>

    Public Function IsUp(Url As Uri) As Boolean
        Return Url.AbsoluteUri.IsUp()
    End Function


















End Module

File: InnerLibs/Web.cs
using InnerLibs.Console;
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.RegularExpressions;

namespace InnerLibs
{
    /// <summary>
    /// Modulo WebExtensions
    /// </summary>
    /// <remarks></remarks>
    public static partial class Util
    {
        #region Public Methods

        /// <summary>
        /// Adciona um parametro a Query String de uma URL
        /// </summary>
        /// <param name="Url">Uri</param>
        /// <param name="Key">Nome do parâmetro</param>
        /// <param name="Values">Valor do Parâmetro</param>
        /// <returns></returns>
        public static Uri AddParameter(this Uri Url, string Key, bool Append, params string[] Values)
        {
            var UriBuilder = new UriBuilder(Url);
            var query = UriBuilder.Query.ParseQueryString();
            if (Values is null || Append == false)
            {
                if (query.AllKeys.Contains(Key))
                {
                    query.Remove(Key);
                }
            }

            foreach (var v in Values ?? Array.Empty<string>())
            {
                query.Add(Key, v);
            }

            UriBuilder.Query = query.ToString();
            Url = new Uri(UriBuilder.ToString());
            return Url;
        }

        /// <summary>
        /// Adciona um parametro a Query String de uma URL
        /// </summary>
        /// <param name="Url">Uri</param>
        /// <param name="Key">Nome do parâmetro</param>
        /// <param name="Values">Valor do Parâmetro</param>
        /// <returns></returns>
        public static Uri AddParameter(this Uri Url, string Key, params string[] Values) => Url.AddParameter(Key, true, Values);

        public static byte[] DownloadFile(string URL, NameValueCollection Headers = null, Encoding Encoding = null)
        {
            byte[] s = null;
            using (var c = new WebClient())
            {
                c.Encoding = Encoding ?? new UTF8Encoding(false);

                if (Headers != null)
                    c.Headers.Add(Headers);
                if (URL.IsURL()) s = c.DownloadData(URL);
            }

            return s;
        }

        public static byte[] DownloadFile(this Uri URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadFile($"{URL}", Headers, Encoding);

        public static System.Drawing.Image DownloadImage(string URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadFile(URL, Headers, Encoding).ToImage();

        public static System.Drawing.Image DownloadImage(this Uri URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadImage($"{URL}", Headers, Encoding);

        public static T DownloadJson<T>(string URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadString(URL, Headers, Encoding).FromJson<T>();

        public static object DownloadJson(string URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadString(URL, Headers, Encoding).FromJson();

        public static T DownloadJson<T>(this Uri URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadJson<T>($"{URL}", Headers, Encoding);

        public static object DownloadJson(this Uri URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadJson($"{URL}", Headers, Encoding);

        public static string DownloadString(string URL, NameValueCollection Headers = null, Encoding Encoding = null)
        {
            string s = Util.EmptyString;
            using (var c = new WebClient())
            {
                c.Encoding = Encoding ?? new UTF8Encoding(false);

                if (Headers != null)
                    c.Headers.Add(Headers);

                if (URL.IsURL()) s = $"{c.DownloadString(URL)}";
            }

            return s;
        }

        public static string DownloadString(this Uri URL, NameValueCollection Headers = null, Encoding Encoding = null) => DownloadString($"{URL}", Headers, Encoding);

        /// <summary>
        /// Retorna o Titulo do arquivo a partir do nome do arquivo
        /// </summary>
        /// <param name="Info">Arquivo ou Diretório</param>
        /// <returns></returns>
        public static string FileNameAsTitle(this FileSystemInfo Info, bool ForceCase = false) => Path.GetFileNameWithoutExtension(Info?.Name).ToNormalCase().ToTitle(ForceCase);

        /// <summary>
        /// Retorna o Titulo do arquivo a partir do nome do arquivo
        /// </summary>
        /// <param name="FilePath">Arquivo ou Diretório</param>
        /// <returns></returns>
        public static string FileNameAsTitle(this string FilePath, bool ForceCase = false) => Path.GetFileNameWithoutExtension(FilePath).ToNormalCase().ToTitle(ForceCase);

        /// <summary>
        /// Captura o Username ou UserID de uma URL do Facebook
        /// </summary>
        /// <param name="URL">URL do Facebook</param>
        /// <returns></returns>
        public static string GetFacebookUsername(this string URL) => URL.IsURL() && URL.GetDomain().ToLowerInvariant().IsAny("facebook.com", "fb.com")
               ? Regex.Match(URL.Replace("fb.com", "facebook.com"), @"(?:(?:http|https):\/\/)?(?:www.)?facebook.com\/(?:(?:\w)*#!\/)?(?:pages\/)?(?:[?\w\-]*\/)?(?:profile.php\?id=(?=\d.*))?([\w\-]*)?").Groups[1].Value
                 : throw new ArgumentException("Invalid Facebook URL", nameof(URL));

        /// <summary>
        /// Captura o Username ou UserID de uma URL do Facebook
        /// </summary>
        /// <param name="URL">URL do Facebook</param>
        /// <returns></returns>
        public static string GetFacebookUsername(this Uri URL) => URL?.AbsoluteUri.GetFacebookUsername();

        public static string GetFileNameWithoutExtension(this FileInfo Info) => Info != null ? Path.GetFileNameWithoutExtension(Info.Name) : Util.EmptyString;

        /// <summary>
        /// Retorna um novo <see cref="FileInfo"/> substituindo a extensão original por <paramref name="Extension"/>
        /// </summary>
        /// <param name="Info"></param>
        /// <param name="Extension"></param>
        /// <returns></returns>
        public static FileInfo ReplaceExtension(this FileInfo Info, string Extension)
        {
            if (Info != null)
                return new FileInfo(Path.Combine(Info.DirectoryName, Info.GetFileNameWithoutExtension() + "." + Extension.IfBlank("bin").TrimStart('.')).FixPath());
            return null;
        }

        public static IEnumerable<string> GetIPs() => GetLocalIP().Union(new[] { GetPublicIP() });

        public static IEnumerable<string> GetLocalIP()
        {
            var host = Dns.GetHostEntry(Dns.GetHostName());
            foreach (var ip in host.AddressList)
            {
                if (ip.AddressFamily == AddressFamily.InterNetwork)
                {
                    yield return ip.ToString().Trim();
                }
            }
        }

        public static string GetPublicIP()
        {
            var IP = InnerLibs.Util.EmptyString;
            Util.TryExecute(() => IP = DownloadString("https://ipv4.icanhazip.com/")).ConsoleWriteError();
            IP = IP.Trim().NullIf(x => !x.IsIP());
            return IP.Trim();
        }

        /// <summary>
        /// Retorna os segmentos de uma url
        /// </summary>
        /// <param name="URL"></param>
        /// <returns></returns>
        public static IEnumerable<string> GetURLSegments(this string URL)
        {
            var l = new List<string>();
            var p = new Regex(@"(?<!\?.+)(?<=\/)[\w-.]+(?=[/\r\n?]|$)", (RegexOptions)((int)RegexOptions.Singleline + (int)RegexOptions.IgnoreCase));
            var gs = p.Matches(URL);
            foreach (Match g in gs)
            {
                l.Add(g.Value);
            }

            return l;
        }

        /// <summary>
        /// Captura o ID de um video do YOUTUBE ou VIMEO em uma URL
        /// </summary>
        /// <param name="URL">URL do video</param>
        /// <returns>ID do video do youtube ou Vimeo</returns>
        public static string GetVideoID(this Uri URL) => GetVideoID(URL.AbsoluteUri);

        /// <summary>
        /// Captura o ID de um video do youtube em uma URL
        /// </summary>
        /// <param name="URL">URL do video</param>
        /// <returns>ID do video do youtube</returns>
        public static string GetVideoID(this string URL)
        {
            if (URL.IsURL())
            {
                if (URL.GetDomain().ContainsAny("youtube", "youtu"))
                {
                    return Regex.Match(URL.ReplaceNone("&feature=youtu.be"), @"(?:https?:\/\/)?(?:www\.)?youtu(?:.be\/|be\.com\/watch\?v=|be\.com\/v\/)(.{8,})").Groups[1].Value;
                }
                else if (URL.GetDomain().ContainsAny("vimeo"))
                {
                    return Regex.Match(URL, @"vimeo\.com/(?:.*#|.*/videos/)?([0-9]+)").Groups[1].Value;
                }
            }

            throw new ArgumentException("Invalid Youtube or Vimeo URL", nameof(URL));
        }

        /// <summary>
        /// Captura a Thumbnail de um video do youtube
        /// </summary>
        /// <param name="URL">Url do Youtube</param>
        /// <returns></returns>
        public static byte[] GetYoutubeThumbnail(string URL) => DownloadFile($"http://img.youtube.com/vi/{GetVideoID(URL)}/hqdefault.jpg");

        /// <summary>
        /// Captura a Thumbnail de um video do youtube
        /// </summary>
        /// <param name="URL">Url do Youtube</param>
        /// <returns></returns>
        public static byte[] GetYoutubeThumbnail(this Uri URL) => GetYoutubeThumbnail(URL?.AbsoluteUri);

        /// <summary>
        /// Verifica se o computador está conectado com a internet
        /// </summary>
        /// <returns></returns>
        public static bool IsConnected(string Test = "http://google.com")
        {
            try
            {
                using (var client = new WebClient())
                {
                    using (var stream = client.OpenRead(Test))
                    {
                        return true;
                    }
                }
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Minifica uma folha de estilo CSS
        /// </summary>
        /// <param name="CSS">String contendo o CSS</param>
        /// <returns></returns>
        public static string MinifyCSS(this string CSS, bool PreserveComments = false)
        {
            if (CSS.IsNotBlank())
            {
                CSS = Regex.Replace(CSS, "[a-zA-Z]+#", "#");
                CSS = Regex.Replace(CSS, @"[\n\r]+\s*", InnerLibs.Util.EmptyString);
                CSS = Regex.Replace(CSS, @"\s+", " ");
                CSS = Regex.Replace(CSS, @"\s?([:,;{}])\s?", "$1");
                CSS = CSS.Replace(";}", "}");
                CSS = Regex.Replace(CSS, @"([\s:]0)(px|pt|%|em)", "$1");
                // Remove comments from CSS
                if (PreserveComments == false)
                {
                    CSS = Regex.Replace(CSS, @"/\*[\d\D]*?\*/", InnerLibs.Util.EmptyString);
                }
            }

            return CSS;
        }

        public static NameValueCollection ParseQueryString(this Uri URL) => URL?.Query.ParseQueryString();

        /// <summary>
        /// Remove um parametro da Query String de uma URL
        /// </summary>
        /// <param name="Url">Uri</param>
        /// <param name="Key">Nome do parâmetro</param>
        /// <param name="Values">Valor do Parâmetro</param>
        /// <returns></returns>
        public static Uri RemoveParameter(this Uri Url, params string[] Keys)
        {
            var UriBuilder = new UriBuilder(Url);
            var query = UriBuilder.Query.ParseQueryString();
            Keys = Keys != null && Keys.Any() ? Keys : query.AllKeys;
            foreach (var k in Keys)
            {
                try
                {
                    query.Remove(k);
                }
                catch
                {
                }
            }

            UriBuilder.Query = query.ToQueryString();
            return UriBuilder.Uri;
        }

        public static string RemoveUrlParameters(this string URL)
        {
            if ((URL.IsURL()))
            {
                URL = Regex.Replace(URL, @"{([^:]+)\s*:\s*(.+?)(?<!\\)}", InnerLibs.Util.EmptyString);
                URL = URL.TrimLastEqual("/");
            }
            return URL;
        }

        public static string RemoveUrlParameters(Uri URL) => RemoveUrlParameters(URL?.ToString());

        /// <summary>
        /// Substitui os parametros de rota de uma URL por valores de um objeto
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="obj"></param>
        /// <param name="URL"></param>
        /// <returns></returns>
        public static string ReplaceUrlParameters<T>(this string URL, T obj)
        {
            if (URL.IsURL())
            {
                URL = Regex.Replace(URL, @"{([^:]+)\s*:\s*(.+?)(?<!\\)}", "{$1}");
                if (obj != null)
                {
                    URL = URL.Inject(obj);
                }

                URL = URL.TrimLastEqual("/");
            }
            return URL;
        }

        public static string ReplaceUrlParameters<T>(Uri URL, T obj) => ReplaceUrlParameters(URL?.ToString(), obj);

        #endregion Public Methods
    }
}